server/src/api/roadmaps/roadmaps.controller.ts

```diff
@@ -96,3 +96,34 @@ export const deleteRoadmaps: RouteHandlerFnc = async (ctx) => {
 
   ctx.status = numDeleted == 1 ? 200 : 404;
 };
+
+export const leaveRoadmap: RouteHandlerFnc = async (ctx) => {
+  if (!ctx.state.user) throw new Error('User is required');
+  await Role.transaction(async (trx) => {
+    const userRole = await Role.query(trx)
+      .where('userId', ctx.state.user!.id)
+      .andWhere('roadmapId', ctx.params.roadmapId)
+      .first();
+
+    if (userRole.type === RoleType.Admin) {
+      const numAdmins = await Role.query(trx)
+        .where('roadmapId', ctx.params.roadmapId)
+        .andWhere('type', RoleType.Admin)
+        .resultSize();
+
+      if (numAdmins === 1) {
+        ctx.status = 403;
+        ctx.body =
+          'Cannot leave as the last admin of roadmap, delete the whole roadmap instead.';
+        return;
+      }
+    }
+
+    const numDeleted = await Role.query(trx).deleteById([
+      Number(ctx.state.user!.id),
+      Number(ctx.params.roadmapId),
+    ]);
```

Comment:
```suggestion
    const numDeleted = await role.$query(trx).delete();
```

This would avoid duplicating the query criteria from above.